### PR TITLE
[Asan] Provide TTI hook to provide memory reference infromation of target intrinsics.

### PR DESCRIPTION
**Pull request by yetingk**
Previously asan considers target intrinsics as black boxes, so asan could not instrument accurate check. This patch provide TTI hooks to make targets describe their intrinsic informations to asan.

Note,
1. this patch renames InterestingMemoryOperand to MemoryRefInfo.
2. this patch does not support RVV indexed/segment load/store.

### CODE REVIEW

File: llvm/include/llvm/Analysis/MemoryRefInfo.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines MemoryRefInfo class that is used when getting
+// the information of a memory reference instruction.
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_SUPPORT_MEMORYREFINFO_H
```
Comment (topperc) [inline llvm/include/llvm/Analysis/MemoryRefInfo.h:13]: This file is Analysis, this should be `LLVM_ANALYSIS_MEMORYREFINFO_H`

File: llvm/include/llvm/Analysis/TargetTransformInfo.h [hunk 2]
```diff
@@ -955,6 +956,10 @@ class TargetTransformInfo {
   MemCmpExpansionOptions enableMemCmpExpansion(bool OptSize,
                                                bool IsZeroCmp) const;
 
+  // Add MemoryRefInfo of Intrinsic \p II into array \p Interesting.
+  bool getMemoryRefInfo(SmallVectorImpl<MemoryRefInfo> &Interesting,
+                        IntrinsicInst *II) const;
```
Comment (topperc) [inline llvm/include/llvm/Analysis/TargetTransformInfo.h:961]: `const IntrinsicInst *II`

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -36,6 +38,91 @@ static cl::opt<unsigned> SLPMaxVF(
         "exclusively by SLP vectorizer."),
     cl::Hidden);
 
+bool RISCVTTIImpl::getMemoryRefInfo(SmallVectorImpl<MemoryRefInfo> &Interesting,
+                                    IntrinsicInst *II) const {
+  const DataLayout &DL = getDataLayout();
+  unsigned IntNo = II->getIntrinsicID();
+  LLVMContext &C = II->getContext();
+  IRBuilder<> IB(II);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:46]: `IB` is unused

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -13,7 +13,9 @@
 #include "llvm/CodeGen/BasicTTIImpl.h"
 #include "llvm/CodeGen/CostTable.h"
 #include "llvm/CodeGen/TargetLowering.h"
+#include "llvm/IR/IRBuilder.h"
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:16]: Can this be removed if the unused `IRBuilder` object created in getMemoryRefInfo is removed?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -36,6 +38,91 @@ static cl::opt<unsigned> SLPMaxVF(
         "exclusively by SLP vectorizer."),
     cl::Hidden);
 
+bool RISCVTTIImpl::getMemoryRefInfo(SmallVectorImpl<MemoryRefInfo> &Interesting,
+                                    IntrinsicInst *II) const {
+  const DataLayout &DL = getDataLayout();
+  unsigned IntNo = II->getIntrinsicID();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:44]: `Intrinsic::ID IID = II->getIntrinsicID()`. `IntNo` is used by SelectionDAG.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 6]
```diff
@@ -36,6 +38,91 @@ static cl::opt<unsigned> SLPMaxVF(
...
+    Type *MaskType = Ty->getWithNewType(Type::getInt1Ty(C));
+    Value *Mask = ConstantInt::get(MaskType, 1);
+    if (HasMask)
+      Mask = II->getArgOperand(VLIndex - 1);
+    Value *EVL = II->getArgOperand(VLIndex);
+    Interesting.emplace_back(II, PtrOperandNo, IsWrite, Ty, Alignment, Mask,
+                             EVL, Stride);
+    return true;
+  }
+  case Intrinsic::riscv_masked_strided_load:
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:104]: Has this intrinsic been removed yet?

File: llvm/test/Instrumentation/AddressSanitizer/RISCV/asan-rvv-intrinsics.ll [hunk 7]
```diff
@@ -0,0 +1,2209 @@
...
+  <vscale x 1 x i32>* %1,
+  <vscale x 1 x i1> %2,
+  i64 %3)
+  ret void
+}
+
+declare {<vscale x 1 x i32>,<vscale x 1 x i32>} @llvm.riscv.vlseg2.nxv1i32(<vscale x 1 x i32>,<vscale x 1 x i32>, ptr , i64)
+declare {<vscale x 1 x i32>,<vscale x 1 x i32>} @llvm.riscv.vlseg2.mask.nxv1i32(<vscale x 1 x i32>,<vscale x 1 x i32>, ptr, <vscale x 1 x i1>, i64, i64)
+
+define <vscale x 1 x i32> @test_vlseg2_nxv1i32(ptr align 4 %base, i64 %vl) sanitize_address {
```
Comment (topperc) [inline llvm/test/Instrumentation/AddressSanitizer/RISCV/asan-rvv-intrinsics.ll:173]: Is this just showing we don't support this intrinsic yet?

File: llvm/test/Instrumentation/AddressSanitizer/asan-rvv-intrinsics.ll [hunk 8]
```diff
@@ -0,0 +1,2209 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
Comment (topperc) [inline llvm/test/Instrumentation/AddressSanitizer/asan-rvv-intrinsics.ll:1]: Is this the same file as the one in the RISCV directory?

File: llvm/include/llvm/Analysis/TargetTransformInfo.h (hunk 2 above)
Comment (topperc) [reply llvm/include/llvm/Analysis/TargetTransformInfo.h:961]: ok

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 9]
```diff
@@ -36,6 +37,72 @@ static cl::opt<unsigned> SLPMaxVF(
         "exclusively by SLP vectorizer."),
     cl::Hidden);
 
+bool RISCVTTIImpl::getMemoryRefInfo(SmallVectorImpl<MemoryRefInfo> &Interesting,
+                                    IntrinsicInst *II) const {
+  const DataLayout &DL = getDataLayout();
+  Intrinsic::ID IntNo = II->getIntrinsicID();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:43]: IntNo -> IID

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 10]
```diff
@@ -36,6 +37,72 @@ static cl::opt<unsigned> SLPMaxVF(
...
+    // Use the pointer alignment as the element alignment if the stride is a
+    // multiple of the pointer alignment. Otherwise, the element alignment
+    // should be Align(1).
+    unsigned PointerAlign = Alignment.valueOrOne().value();
+    if (!isa<ConstantInt>(Stride) ||
+        cast<ConstantInt>(Stride)->getZExtValue() % PointerAlign != 0)
+      Alignment = Align(1);
+
+    Type *MaskType = Ty->getWithNewType(Type::getInt1Ty(C));
+    Value *Mask = ConstantInt::get(MaskType, 1);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:94]: ConstantInt::getTrue

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 11]
```diff
@@ -36,6 +37,72 @@ static cl::opt<unsigned> SLPMaxVF(
...
+  case Intrinsic::riscv_vse: {
+    bool IsWrite = II->getType()->isVoidTy();
+    Type *Ty = IsWrite ? II->getArgOperand(0)->getType() : II->getType();
+    const auto *RVVIInfo = RISCVVIntrinsicsTable::getRISCVVIntrinsicInfo(IntNo);
+    unsigned VLIndex = RVVIInfo->VLOperand;
+    unsigned PtrOperandNo = VLIndex - 1 - HasMask;
+    MaybeAlign Alignment =
+        II->getArgOperand(PtrOperandNo)->getPointerAlignment(DL);
+    Type *MaskType = Ty->getWithNewType(Type::getInt1Ty(C));
+    Value *Mask = ConstantInt::get(MaskType, 1);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:62]: ConstantInt::getTrue

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 12]
```diff
@@ -36,6 +37,83 @@ static cl::opt<unsigned> SLPMaxVF(
...
+    const auto *RVVIInfo = RISCVVIntrinsicsTable::getRISCVVIntrinsicInfo(IID);
+    unsigned VLIndex = RVVIInfo->VLOperand;
+    unsigned PtrOperandNo = VLIndex - 2 - HasMask;
+    MaybeAlign Alignment =
+        II->getArgOperand(PtrOperandNo)->getPointerAlignment(DL);
+
+    Value *Stride = II->getArgOperand(PtrOperandNo + 1);
+    // Use the pointer alignment as the element alignment if the stride is a
+    // multiple of the pointer alignment. Otherwise, the element alignment
+    // should be the greatest of common divisor of pointer alignment and stride.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:97]: "greatest of common" -> "greatest common"

**Review by topperc — APPROVED** [review]
LGTM

**Comment by TomWeaver18** [comment]
@yetingk Hello and good morning from the UK.

it looks like the test added in this commit has caused the following buildbots to start failing:

https://lab.llvm.org/buildbot/#/builders/46/builds/2895
https://lab.llvm.org/buildbot/#/builders/154/builds/2431
https://lab.llvm.org/buildbot/#/builders/23/builds/1748

any no doubt many others.

Are you able to take a look and get this fixed up? The bots have been red for a good few hours already.

**Pull request merged**